File: src/main/java/com/hp/jipp/client/PrinterStatus.java
package com.hp.jipp.client;

import com.google.auto.value.AutoValue;
import com.hp.jipp.encoding.AttributeGroup;
import com.hp.jipp.model.Attributes;
import com.hp.jipp.model.PrinterState;

import java.io.IOException;
import java.util.List;

import javax.annotation.Nullable;

// All interesting status fields of a printer
@AutoValue
public abstract class PrinterStatus {

    static PrinterStatus of(AttributeGroup attributes) throws IOException {
        PrinterState state = attributes.getValue(Attributes.PrinterState);
        List<String> reasons = attributes.getValues(Attributes.PrinterStateReasons);
        String message = attributes.getValue(Attributes.PrinterStateMessage);
        if (state == null) throw new IOException("Missing " + Attributes.PrinterState.getName());
        return new AutoValue_PrinterStatus(state, reasons, message);
    }

    public abstract PrinterState getState();

    public abstract List<String> getReasons();

    @Nullable
    public abstract String getMessage();

    public String toString() {
        return "Printer{state=" + getState().getName() +
                (getReasons().isEmpty() ? "" : " r=" + getReasons()) +
                (getMessage() == null ? "" : " m=" + getMessage()) +
                "}";
    }
}

File: src/main/java/com/hp/jipp/client/PrinterStatus.kt
package com.hp.jipp.client

import com.hp.jipp.encoding.AttributeGroup
import com.hp.jipp.model.Attributes
import com.hp.jipp.model.PrinterState
import org.jetbrains.annotations.Nullable

import java.io.IOException

// All interesting status fields of a printer
data class PrinterStatus(val state: PrinterState, val reasons: List<String>, @Nullable val message: String?) {

    override fun toString(): String {
        return "Printer{state=" + state.name +
                (if (reasons.isEmpty()) "" else " r=" + reasons) +
                (if (message == null) "" else " m=" + message) +
                "}"
    }

    companion object {
        @Throws(IOException::class)
        @JvmStatic fun of(attributes: AttributeGroup): PrinterStatus {
            val state = attributes.getValue(Attributes.PrinterState)
            val reasons = attributes.getValues(Attributes.PrinterStateReasons)
            val message = attributes.getValue(Attributes.PrinterStateMessage)
            if (state == null) throw IOException("Missing " + Attributes.PrinterState.name)
            return PrinterStatus(state, reasons, message)
        }
    }
}
